cdap-ui/app/cdap/components/PluginJSONCreator/Create/index.tsx

```diff
@@ -45,161 +40,245 @@ const styles = (): StyleRules => {
+  setFilterToName: undefined,
+  filterToCondition: Map<string, Map<string, string>>(),
+  setFilterToCondition: undefined,
+  filterToShowList: Map<string, List<string>>(),
+  setFilterToShowList: undefined,
+  showToInfo: Map<string, Map<string, string>>(),
+  setShowToInfo: undefined,
+};
 
-  public setWidgetInfo = (widgetInfo: any) => {
-    this.setState({ widgetInfo });
-  };
+// Define initial values for internal states within the entire page
+const appInternalInitialState = {
+  activeStep: 0,
+  setActiveStep: undefined,
+  liveView: false,
+  setLiveView: undefined,
+  JSONStatus: JSONStatusMessage.Normal,
+  setJSONStatus: undefined,
+};
 
-  public setLiveView = (liveView: boolean) => {
-    this.setState({ liveView });
-  };
+export const PluginInfoContext = React.createContext(pluginInfoInitialState);
+export const ConfigurationGroupContext = React.createContext(configurationGroupsInitialState);
+export const WidgetContext = React.createContext(widgetsInitialState);
+export const OutputContext = React.createContext(outputInitialState);
+export const FilterContext = React.createContext(filtersInitialState);
+export const AppInternalContext = React.createContext(appInternalInitialState);
 
-  public setWidgetToAttributes = (widgetToAttributes: any) => {
-    this.setState({ widgetToAttributes });
-  };
+// Components within pluginJSONCreator will be able to use only the contexts that they choose.
+// This will prevent unnecessary re-renderings of the UI.
+export const usePluginInfoState = () => React.useContext(PluginInfoContext);
+export const useConfigurationGroupState = () => React.useContext(ConfigurationGroupContext);
+export const useWidgetState = () => React.useContext(WidgetContext);
+export const useOutputState = () => React.useContext(OutputContext);
+export const useFilterState = () => React.useContext(FilterContext);
+export const useAppInternalState = () => React.useContext(AppInternalContext);
 
-  public setOutputName = (outputName: string) => {
-    this.setState({ outputName });
-  };
+const GlobalStateProvider = ({
+  children,
+  pluginInfoContextValue,
+  configuratioGroupContextValue,
+  widgetContextValue,
+  outputContextValue,
+  filterContextValue,
+  appInternalContextValue,
+}) => {
+  return (
+    <PluginInfoContext.Provider value={pluginInfoContextValue}>
+      <ConfigurationGroupContext.Provider value={configuratioGroupContextValue}>
+        <WidgetContext.Provider value={widgetContextValue}>
+          <OutputContext.Provider value={outputContextValue}>
+            <FilterContext.Provider value={filterContextValue}>
+              <AppInternalContext.Provider value={appInternalContextValue}>
+                {children}
+              </AppInternalContext.Provider>
+            </FilterContext.Provider>
+          </OutputContext.Provider>
+        </WidgetContext.Provider>
+      </ConfigurationGroupContext.Provider>
+    </PluginInfoContext.Provider>
+  );
+};
```

Comment:
is this still needed?

Reply:
We need it for JsonMenu only. I think it's okay to keep the nested providers based on what I see from this link: https://blog.axlight.com/posts/four-patterns-for-global-state-with-react-hooks-context-or-redux/